docs/develop/porting.rst

```diff
@@ -0,0 +1,309 @@
+    #define MICROPY_FLOAT_IMPL          (MICROPY_FLOAT_IMPL_FLOAT)
+
+    // type definitions for the specific machine
+    #define BYTES_PER_WORD (4)
+
+    #define MICROPY_MAKE_POINTER_CALLABLE(p) ((void*)((mp_uint_t)(p) | 1))
+
+    // This port is intended to be 32-bit, but unfortunately, int32_t for
+    // different targets may be defined in different ways - either as int
+    // or as long. This requires different printf formatting specifiers
+    // to print such value. So, we avoid int32_t and use int directly.
+    #define UINT_FMT "%u"
+    #define INT_FMT "%d"
+    typedef int mp_int_t; // must be pointer size
+    typedef unsigned mp_uint_t; // must be pointer size
+
+    typedef long mp_off_t;
+
+    #define MP_PLAT_PRINT_STRN(str, len) mp_hal_stdout_tx_strn_cooked(str, len)
+
+    // extra built in names to add to the global namespace
+    #define MICROPY_PORT_BUILTINS \
+        { MP_OBJ_NEW_QSTR(MP_QSTR_open), (mp_obj_t)&mp_builtin_open_obj },
+
+    // We need to provide a declaration/definition of alloca()
+    #include <alloca.h>
+
+    #define MICROPY_HW_BOARD_NAME "example-board""
+    #define MICROPY_HW_MCU_NAME   "unknown-cpu"
+
+    #ifdef __linux__
+    #define MICROPY_MIN_USE_STDOUT (1)
+    #endif
+
+    #ifdef __thumb__
+    #define MICROPY_MIN_USE_CORTEX_CPU (1)
+    #define MICROPY_MIN_USE_STM32_MCU (1)
+    #endif
+
+    #define MP_STATE_PORT MP_STATE_VM
+
+    #define MICROPY_PORT_ROOT_POINTERS \
+        const char *readline_hist[8];
+
+    #endif
+
+This configuration file contains machine-specific configurations including aspects like if different
+MicroPython features should be enabled i.e ``#define MICROPY_ENABLE_GC (1) ``. Making this Setting
+``0`` disables the feature.
+
+Other configurations include type definitions, root pointers, board name, microcontroller name
+etc.
+
+Similarly, an example ``mphalport.h`` file looks like this:
+
+.. code-block:: c
+
+    static inline mp_uint_t mp_hal_ticks_ms(void) { return 0; }
+    static inline void mp_hal_set_interrupt_char(char c) {}
+
+Support for standard input/output
+----------------------------------
+
+Now add a file ``uart_core.c`` to allow for input and output:
+
+.. code-block:: c
+
+    #include <unistd.h>
+    #include "py/mpconfig.h"
+
+
+    // Receive single character
+    int mp_hal_stdin_rx_chr(void) {
+        unsigned char c = 0;
+        #if MICROPY_MIN_USE_STDOUT
+        int r = read(STDIN_FILENO, &c, 1);
+        (void)r;
+        #endif
+        return c;
+    }
+
+    // Send string of given length
+    void mp_hal_stdout_tx_strn(const char *str, mp_uint_t len) {
+        #if MICROPY_MIN_USE_STDOUT
+        int r = write(STDOUT_FILENO, str, len);
+        (void)r;
+        #endif
+    }
+
+These input and output functions have to be modified depending on the
+specific board API. This example uses the standard input/output stream.
+
+Adding a module to a port
+--------------------------
+
```

Comment:
I struggled a lot with this section but eventually got it working. Following this section I couldn't get a module to import. Through a lot of trial & error, I eventually came up with a procedure that worked consistently with “extended modules” (EXTMOD) & local Port modules. Further testing (by others too?) may see a refinement on this procedure.

What would be really useful is to turn this section into a step-wise procedure that guarantees a positive result. A checklist could also be added to assure success.

This is the procedure I came up with: **Example objective:  To add the ‘extended module’ ‘modujson’**:

- `cd extmod/`

- Open **“modujson.c”** in an editor.
- Near the top of the file, after the “includes”, look for a macro **“#if MICROPY_PY_*”**. This is used in the **“mpconfigport.h”** as a module ‘enabling’ flag. Copy this name:
- `#if MICROPY_PY_UJSON`
- Find at the end of ‘**modujson.c**’ the phrase “**const mp_obj_module_t**”.  Copy the name given as the next attribute. Eg:
- “**`const mp_obj_module_t mp_module_ujson = {`**“
- This is needed in the Port’s “**mpconfigport.h**” file.
- Returning to your Port:
- `cd ports/samd`
- Edit “`mpconfigport.h`”
- At the top of the file, add to the list of definitions, defining the directive for the module and enabling:
- `#define MICROPY_PY_UJSON                    (1)`
- Declare the external variable as defined in the module’s C file (with other declarations):
- **`extern const struct _mp_obj_module_t mp_module_ujson;`**
- Add the module to the “unique string” (**QSTR**) declarations:

```
 #define MICROPY_PORT_BUILTIN_MODULES \
     { MP_ROM_QSTR(MP_QSTR_machine), MP_ROM_PTR(&mp_module_machine) }, \
     { MP_ROM_QSTR(MP_QSTR_utime), MP_ROM_PTR(&mp_module_utime) }, \
     { MP_ROM_QSTR(MP_QSTR_ujson), MP_ROM_PTR(&mp_module_ujson) }, \

```

 *****************************************************************
 Question for the Developers: I’ve seen 2 conventions for this part of the file:
 Following existing convention in the Port’s mpconfigport.h file:
 `{ MP_ROM_QSTR(MP_QSTR_ujson), MP_ROM_PTR(&mp_module_ujson) }, \`
 Following existing convention in the document’s mpconfigport.h example:
 `{ MP_OBJ_NEW_QSTR(MP_QSTR_ujson), (mp_obj_t)&mp_module_ujson}, \`
 Both seem to work, but why the difference and what convention is optimal?
 *****************************************************************
- Edit ‘**Makefile**’ to add the module’s c file to the list for compilation. This seems to be the convention:
```
 EXTMOD_SRC_C = $(addprefix extmod/,\
     modure.c \
     modujson.c \
     )
```
- **Not ‘modujson**’, but if the module is local & unique to the Port (ie; in the Port’s directory, eg; “utime”), add the ‘C’ file to the **SRC_QSTR**:
```
 # List of sources for qstr extraction for c modules in Port dir.
 SRC_QSTR += modutime.c modmachine.c moduos.c
```
 
- Compile MicroPython and attempt to fix errors. My experience is that most errors relate to incorrectly defined directories:
```
make clean
make
```